File: src/Test/ArrayOpsFFI.php
<?php
$exports['runArrayOpsFFI'] = function($limit) {
    $n = (int)$limit;
    
    $range = function($start) {
        return function($end) use ($start) {
            $arr = [];
            $step = $end >= $start ? 1 : -1;
            for ($i = $start; $step > 0 ? $i <= $end : $i >= $end; $i += $step) {
                $arr[] = $i;
            }
            return $arr;
        };
    };

    $filter = function($p) {
        return function($arr) use ($p) {
            $res = [];
            foreach ($arr as $x) {
                if ($p($x)) {
                    $res[] = $x;
                }
            }
            return $res;
        };
    };

    $foldl = function($f, $initial, $arr) {
        $acc = $initial;
        foreach ($arr as $x) {
            $acc = $f($acc)($x);
        }
        return $acc;
    };

    $rng = $range(1)($n);
    $filtered = $filter(function($x) { return $x % 2 === 0; })($rng);
    return $foldl(function($acc) { return function($x) use ($acc) { return $acc + $x; }; }, 0, $filtered);
};
return $exports;

File: src/Test/ArrayOpsFFI.cc
#include "purescript.h"
#include <vector>

namespace {
    std::vector<int> range(int start, int end) {
        std::vector<int> values;
        const int step = start <= end ? 1 : -1;
        for (int value = start;; value += step) {
            values.push_back(value);
            if (value == end) return values;
        }
    }

    template <class A, class Predicate>
    std::vector<A> filter(Predicate predicate, const std::vector<A>& values) {
        std::vector<A> accepted;
        for (const A& value : values) {
            if (predicate(value)) accepted.push_back(value);
        }
        return accepted;
    }

    template <class A, class B, class Combine>
    B foldl(Combine combine, B acc, const std::vector<A>& values) {
        for (const A& value : values) acc = combine(acc, value);
        return acc;
    }
}

FOREIGN_BEGIN(Test_ArrayOpsFFI)
exports["runArrayOpsFFI"] = [](const boxed& in) -> boxed {
    const auto values = range(1, unbox<int>(in));
    const auto evens = filter([](int value) { return value % 2 == 0; }, values);
    return foldl([](int sum, int value) { return sum + value; }, 0, evens);
};
FOREIGN_END
